drivers/clock_control/nrf_power_clock.c

```diff
@@ -1,280 +1,333 @@
+{
+	bool ret = nrf_power_event_check(evt) &&
+			nrf_power_int_enable_check(intmask);
 
-	if (!(intenset & CLOCK_INTENSET_LFCLKSTARTED_Msk)) {
-		nrf_clock_int_disable(NRF_CLOCK_INT_LF_STARTED_MASK);
+	if (ret) {
+		nrf_power_event_clear(evt);
 	}
 
-	NVIC_ClearPendingIRQ(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-
-	irq_enable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
+	return ret;
+}
+#endif
 
-#else /* !CONFIG_CLOCK_CONTROL_NRF_K32SRC_BLOCKING */
-	/* NOTE: LFCLK will initially start running from the LFRC if LFXO is
-	 *       selected.
-	 */
-	nrf_clock_int_enable(NRF_CLOCK_INT_LF_STARTED_MASK);
-	nrf_clock_task_trigger(NRF_CLOCK_TASK_LFCLKSTART);
-#endif /* !CONFIG_CLOCK_CONTROL_NRF_K32SRC_BLOCKING */
-
-#if NRF_CLOCK_HAS_CALIBRATION
-	/* If RC selected, calibrate and start timer for consecutive
-	 * calibrations.
-	 */
-	nrf_clock_int_disable(NRF_CLOCK_INT_DONE_MASK |
-			      NRF_CLOCK_INT_CTTO_MASK);
-	NRF_CLOCK->EVENTS_DONE = 0;
-	NRF_CLOCK->EVENTS_CTTO = 0;
-
-	if ((lf_clk_src & CLOCK_LFCLKSRC_SRC_Msk) == CLOCK_LFCLKSRC_SRC_RC) {
-		int err;
-
-		/* Set the Calibration Timer Initial Value */
-		NRF_CLOCK->CTIV = 16;	/* 4s in 0.25s units */
-
-		/* Enable DONE and CTTO IRQs */
-		nrf_clock_int_enable(NRF_CLOCK_INT_DONE_MASK |
-				     NRF_CLOCK_INT_CTTO_MASK);
-
-		/* If non-blocking LF clock start, then start HF clock in ISR */
-		if ((NRF_CLOCK->LFCLKSTAT & CLOCK_LFCLKSTAT_STATE_Msk) == 0) {
-			nrf_clock_int_enable(NRF_CLOCK_INT_LF_STARTED_MASK);
-			goto lf_already_started;
-		}
+static void usb_power_isr(void)
+{
+#if defined(CONFIG_USB_NRF52840)
+	extern void usb_dc_nrfx_power_event_callback(nrf_power_event_t event);
 
-		/* Start HF clock, if already started then explicitly
-		 * assert IRQ.
-		 * NOTE: The INTENSET is used as state flag to start
-		 * calibration in ISR.
-		 */
-		nrf_clock_int_enable(NRF_CLOCK_INT_HF_STARTED_MASK);
+	if (power_event_check_and_clean(NRF_POWER_EVENT_USBDETECTED,
+					NRF_POWER_INT_USBDETECTED_MASK)) {
+		usb_dc_nrfx_power_event_callback(NRF_POWER_EVENT_USBDETECTED);
+	}
 
-		err = m16src_start(dev, false);
-		if (!err) {
-			NVIC_SetPendingIRQ(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-		} else {
-			__ASSERT_NO_MSG(err == -EINPROGRESS);
-		}
+	if (power_event_check_and_clean(NRF_POWER_EVENT_USBPWRRDY,
+					NRF_POWER_INT_USBPWRRDY_MASK)) {
+		usb_dc_nrfx_power_event_callback(NRF_POWER_EVENT_USBPWRRDY);
 	}
-#endif /* NRF_CLOCK_HAS_CALIBRATION */
 
-lf_already_started:
-	stat = (NRF_CLOCK->LFCLKSRCCOPY & CLOCK_LFCLKSRCCOPY_SRC_Msk) |
-	       CLOCK_LFCLKSTAT_STATE_Msk;
-	if ((NRF_CLOCK->LFCLKSTAT & stat) == stat) {
-		return 0;
-	} else {
-		return -EINPROGRESS;
+	if (power_event_check_and_clean(NRF_POWER_EVENT_USBREMOVED,
+					NRF_POWER_INT_USBREMOVED_MASK)) {
+		usb_dc_nrfx_power_event_callback(NRF_POWER_EVENT_USBREMOVED);
 	}
+#endif
 }
 
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-static inline void power_event_cb(nrf_power_event_t event)
+static void pof_isr(void)
```

Comment:
Do we actually need this? The `POFWARN` interrupt is not active, so this function actually does nothing but a little delay.

Reply:
right, when POF support will be added then it makes sense. removed.